### PR TITLE
Cuda recipe

### PR DESCRIPTION
**Pull request by andyneff**
Recipe to install CUDA into a non-CUDA container

### CODE REVIEW

File: recipe_cuda.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+              parse_envvar "${1}/runtime/Dockerfile" > "/usr/local/share/just/info/cuda/12_cudaenv_${2}"; \
+              parse_envvar "${1}/devel/Dockerfile" > "/usr/local/share/just/info/cuda/14_cudaenv_${2}"; \
+              if [ -n "${CUDNN_VERSION:+set}" ]; then \
+                # If this errors during build, you probably set CUDNN_VERSION wrong
+                parse_envvar "${1}/runtime/cudnn${CUDNN_VERSION}/Dockerfile" > "/usr/local/share/just/info/cuda/13_cudaenv_${2}"; \
+                parse_envvar "${1}/devel/cudnn${CUDNN_VERSION}/Dockerfile" > "/usr/local/share/just/info/cuda/13_cudaenv_${2}"; \
+              fi; \
+            }; \
+            # Env Vars
+            # Currently all the fedoras use the same vars, so the first one is good enough
```
Comment (scott-vsi) [inline recipe_cuda.Dockerfile:63]: what does that mean, "the first one"

File: recipe_cuda.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            done; \
+            # Ubuntus
+            for ubuntu in ubuntu2804 ubuntu2604 ubuntu2404 ubuntu2204 ubuntu2004 ubuntu1804; do \
+              if [ -d "/cuda/dist/${CUDA_VERSION}/${ubuntu}" ]; then \
+                parse_os "/cuda/dist/${CUDA_VERSION}/${ubuntu}" ubuntu; \
+                break; \
+              fi; \
+            done; \
+            # get GPG keys
+            rhel_keys=($(grep -hr 'rhel.*\.pub' /cuda/dist/[0-9]* | sed -E 's|.*(https://[^ ]*).*|\1|' | sort -u)); \
```
Comment (scott-vsi) [inline recipe_cuda.Dockerfile:78]: what directory are we in here?

File: README.rst [hunk 3]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+
    RUN dnf install -y --enablerepo=rocky-appstream telnet # This line is just an example
    ...
 
+CUDA
+----
+
+=========================== ============
+Name                        CUDA
+Build Args                  ``CUDA_REPO_REF`` - The version of the CUDA container repo parsed
```
Comment (scott-vsi) [inline README.rst:346]:
```suggestion
Build Args                  ``CUDA_REPO_REF`` - SHA of the CUDA container repo parsed
```

File: README.rst [hunk 4]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+Build Args                  ``CUDA_VERSION`` - Version of CUDA to install (e.g. ``10.2`` or ``11.0.7``)
+Build Args                  ``CUDNN_VERSION`` - Optional: Version of CUDNN to install. (e.g. ``7`` or ``8``)
+Build Args                  ``CUDA_RECIPE_TARGET`` - Optional: Sets the target level of CUDA install (explained below). Default: ``runtime``
+Environment Variable        ``NVIDIA_VISIBLE_DEVICES`` - Required: Sets while nvidia devices to connect. Set to ``all``
```
Comment (scott-vsi) [inline README.rst:350]:
```suggestion
Environment Variable        ``NVIDIA_VISIBLE_DEVICES`` - Required: Sets which nvidia devices are visible in the container. Default: ``all``
```

File: README.rst [hunk 5]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+Environment Variable        ``NVIDIA_DRIVER_CAPABILITIES`` - Optional: Default is ``compute,utility`` when unset
```
Comment (scott-vsi) [inline README.rst:351]:
```suggestion
Environment Variable        ``NVIDIA_DRIVER_CAPABILITIES`` - Optional: Which device capabilities are enabled in the container. Default (when unset): ``compute,utility``
```

File: README.rst [hunk 6]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+Environment Variable        ``NVIDIA_REQUIRE_*`` - Optional: Sets test conditions to prevent running on incompatible systems
+Output dir                  ``/usr/local``
+Minimum Dockerfile frontend docker/dockerfile:1.3-labs or docker/dockerfile:1.4
```
Comment (scott-vsi) [inline README.rst:354]:
```suggestion
Minimum Dockerfile frontend: docker/dockerfile:1.3-labs or docker/dockerfile:1.4
```

File: README.rst [hunk 7]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+=========================== ============
```
Comment (scott-vsi) [inline README.rst:355]:
```suggestion
=======================================
```

File: README.rst [hunk 8]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+
+While starting from a base image with CUDA already setup for docker is ideal, when we have to be based on a specific image (e.g. hardened images), this becomes impossible or impractical. Instead we need to start with a particular image and add CUDA support to it.
+
+Currently only RHEL and Ubuntu based images are supported (not Fedora or Debian).
```
Comment (scott-vsi) [inline README.rst:359]:
```suggestion
Currently, only RHEL and Ubuntu based images are supported (not Fedora or Debian).
```

File: recipe_cuda.Dockerfile (hunk 1 above)
Comment (scott-vsi) [reply recipe_cuda.Dockerfile:63]: you mention that Fedora is not supported in the REAME. should this say all redhats

File: recipe_cudagl.Dockerfile [hunk 9]
```diff
@@ -0,0 +1,133 @@
+# syntax=docker/dockerfile:1.4
+
+FROM scratch as scripts
+
+COPY --chmod=755 <<EOF /setup
```
Comment (scott-vsi) [inline recipe_cudagl.Dockerfile:5]: uhhh. I did not realize you could execute arbitrary code in a COPY. what does this do with the files it builds?

File: recipe_cudagl.Dockerfile [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+              # Headers part 3
+              git clone --branch=mesa-17.3.3 --depth=1 https://gitlab.freedesktop.org/mesa/mesa.git; \
+              cd mesa; \
+                mkdir /usr/local/include/GL; \
+                cp include/GL/gl.h include/GL/gl_mangle.h /usr/local/include/GL/; \
+              cd ..; \
+              # cleanup
+              apk del .del; \
+              mkdir /usr/local/share/just/info/cuda/; \
+              echo 'PKG_CONFIG_PATH=${PKG_CONFIG_PATH-}${PKG_CONFIG_PATH:+:}/usr/local/lib64/pkgconfig:/usr/local/lib/pkgconfig' > /usr/local/share/just/info/cuda/30_pkgconfig_rhel ;;\
```
Comment (scott-vsi) [inline recipe_cudagl.Dockerfile:125]: 2 semicolons?

File: README.rst [hunk 11]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+
+There are many steps to setting up CUDA in an image:
+
+- Setting up CUDA repo and GPG key
```
Comment (scott-vsi) [inline README.rst:363]:
```suggestion
- Setting up the CUDA repo GPG key
```

File: README.rst [hunk 12]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+- Installing the right packages, so that we limit the amount of bloat to the image
+- Setting various environment variables at container run time
+- Setting certain environment variables at container create time
```
Comment (scott-vsi) [inline README.rst:366]: what is the difference between these to cases? should this say 'container build time'? (mentioned in the next sections too)

File: README.rst [hunk 13]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+Environment variables
+---------------------
+
+Because of how the nvidia runtime operates, it needs certain variables set at container create time. Some ways this can be accomplished include:
+
+- Adding an ``ENV`` to the Dockerfile
+- Adding to the ``environment:`` section in ``docker-compose.yml``
+- Adding the ``-e`` flag to the ``docker run`` call
+
+``NVIDIA_VISIBLE_DEVICES`` must be set, or else the nvidia runtime will not activate, and you will have no gpu support. It's suggested to make set ``NVIDIA_VISIBLE_DEVICES`` to ``all`` and allow for an environment variable on the host (i.e. ``local.env``) to override the value to use specific GPUs as needed.
```
Comment (scott-vsi) [inline README.rst:380]:
```suggestion
``NVIDIA_VISIBLE_DEVICES`` must be set, or else the nvidia runtime will not activate, and you will have no GPU support. It's suggested to set ``NVIDIA_VISIBLE_DEVICES`` to ``all`` and allow for an environment variable on the host (i.e. in ``local.env``) to override the value to use specific GPUs as needed.
```

File: README.rst [hunk 14]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+
+.. code-block:: Dockerfile
+
+   # syntax=docker/dockerfile:1.4
+   FROM vsiri/recipe:cuda as cuda
+
+   FROM redhat/ubi8
+   COPY --from=cuda /usr/local /usr/local
+   # Only needs to be run once for all recipes
+   RUN for patch in /usr/local/share/just/container_build_patch/*; do "${patch}"; done
```
Comment (scott-vsi) [inline README.rst:408]: it might be worth noting that if using `just` this is done for you

File: README.rst [hunk 15]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+   ENV NVIDIA_VISIBLE_DEVICES=all # Required for this recipe
+   ENV NVIDIA_DRIVER_CAPABILITIES=compute,utility # Optional: compute,utility is the default if unset
+
+   # (Uncommon) If you need all the nvidia environment variables, source this file
+   RUN source /usr/local/share/just/user_run_patch/10_load_cuda_env; \
+       cmake # This line is just an example
```
Comment (scott-vsi) [inline README.rst:414]:
```suggestion
       cmake . # This line is just an example
```

File: README.rst [hunk 16]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
 
+CUDA
+----
+
+=========================== ============
+Name                        CUDA
+Build Args                  ``CUDA_REPO_REF`` - The version of the CUDA container repo parsed
+Build Args                  ``CUDA_VERSION`` - Version of CUDA to install (e.g. ``10.2`` or ``11.0.7``)
+Build Args                  ``CUDNN_VERSION`` - Optional: Version of CUDNN to install. (e.g. ``7`` or ``8``)
+Build Args                  ``CUDA_RECIPE_TARGET`` - Optional: Sets the target level of CUDA install (explained below). Default: ``runtime``
```
Comment (scott-vsi) [inline README.rst:349]:
```suggestion
Build Args                  ``CUDA_RECIPE_TARGET`` - Optional: Specifies how much of the CUDA stack to install (explained below). Default: ``runtime``
```

File: README.rst [hunk 17]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+
+----------
+Build Args
+----------
+
+The ``CUDA_VERSION``/``CUDNN_VERSION`` build args must be limited to the versions of CUDA in the `nvidia package repos <https://developer.download.nvidia.com/compute/cuda/repos/>`_. Attempting combinations of OS versions and CUDA versions not in the nvidia `codebase <https://gitlab.com/nvidia/container-images/cuda/-/tree/master/dist>`_ will probably fail because those versions of CUDA most likely do not exist in the nvidia package repos. Currently the end-of-life directory is not supported.
+
+``CUDA_RECIPE_TARGET`` is used to specify how much of CUDA you install (devel vs runtime):
```
Comment (scott-vsi) [inline README.rst:392]:
```suggestion
``CUDA_RECIPE_TARGET`` is used to specify how much of the CUDA stack to install (devel vs runtime):
```

File: README.rst [hunk 18]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+   ...
+
+CUDA GL
+-------
+
+=========================== ============
+Name                        CUDA GL
+Build Args                  ``CUDA_RECIPE_TARGET`` - Sets the target level of CUDA install
+Build Args                  ``LIBGLVND_VERSION`` - The version of the GLVND used. Default: ``v1.2.0``
+Environment Variable        ``NVIDIA_DRIVER_CAPABILITIES`` - For opengl offscreen rendering, you at least need `graphics,compute,utility`
```
Comment (scott-vsi) [inline README.rst:424]:
```suggestion
Environment Variable        ``NVIDIA_DRIVER_CAPABILITIES`` - For OpenGL offscreen rendering, you at least need `graphics,compute,utility`
```

File: README.rst [hunk 19]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+Output dir                  ``/usr/local``
+Minimum Dockerfile frontend docker/dockerfile:1.3-labs or docker/dockerfile:1.4
```
Comment (scott-vsi) [inline README.rst:426]:
```suggestion
Minimum Dockerfile frontend: docker/dockerfile:1.3-labs or docker/dockerfile:1.4
```

File: README.rst [hunk 20]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+   RUN source /usr/local/share/just/user_run_patch/10_load_cuda_env; \
+       cmake # This line is just an example
+   ...
+
+CUDA GL
+-------
+
+=========================== ============
+Name                        CUDA GL
+Build Args                  ``CUDA_RECIPE_TARGET`` - Sets the target level of CUDA install
```
Comment (scott-vsi) [inline README.rst:422]:
```suggestion
Build Args                  ``CUDA_RECIPE_TARGET`` - Specifies how much of the CUDA stack to install (explained further above in the CUDA recipe). Default: ``runtime``
```

File: README.rst [hunk 21]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+Build Args                  ``LIBGLVND_VERSION`` - The version of the GLVND used. Default: ``v1.2.0``
+Environment Variable        ``NVIDIA_DRIVER_CAPABILITIES`` - For opengl offscreen rendering, you at least need `graphics,compute,utility`
+Output dir                  ``/usr/local``
+Minimum Dockerfile frontend docker/dockerfile:1.3-labs or docker/dockerfile:1.4
+=========================== ============
+
+Similar to the CUDA recipe, ``CUDA_RECIPE_TARGET`` tells the recipe whether to install runtime or development dependencies. ``devel-only`` still installs both runtime and devel packages, due to how the recipe is structured.
```
Comment (scott-vsi) [inline README.rst:429]:
```suggestion
Similar to the CUDA recipe, ``CUDA_RECIPE_TARGET`` tells the recipe whether to install runtime or development dependencies, although ``devel-only`` installs both runtime and devel packages, due to how the recipe is structured.
```

File: README.rst [hunk 22]
```diff
@@ -334,10 +334,119 @@ This recipe is a little different from other recipes in that it's just a script
...
+You will need to set the environment variable ``NVIDIA_DRIVER_CAPABILITIES`` to allow graphics capabilities which most GL applications will need.
+
+.. rubric:: Example
+
+.. code-block:: Dockerfile
+
+   # syntax=docker/dockerfile:1.4
+   FROM vsiri/recipe:cudagl as cudagl
+
+   FROM some_cuda_image
```
Comment (scott-vsi) [inline README.rst:442]: you use a real image above. I would do the same here

File: recipe_cudagl.Dockerfile [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+    }
+}
+EOF
+
+ONBUILD ARG CUDA_RECIPE_TARGET=runtime
+
+ONBUILD RUN case "${CUDA_RECIPE_TARGET}" in *devel*) \
+              apk add --no-cache --virtual .del git; \
+              # Headers part 1
+              git clone https://github.com/KhronosGroup/OpenGL-Registry.git; \
```
Comment (scott-vsi) [inline recipe_cudagl.Dockerfile:105]: consider just curling the REF you need instead of cloning the whole repo

File: README.rst (hunk 5 above)
Comment (andyneff) [reply README.rst:351]:
```suggestion
Environment Variable        ``NVIDIA_DRIVER_CAPABILITIES`` - Optional: Which device capabilities are enabled in the container. Default: `compute,utility`, which is the meaning by the runtime if this environment variable is unset.
```

File: README.rst (hunk 3 above)
Comment (andyneff) [reply README.rst:346]: Doesn't have to be a sha, it's a ref

File: 30_install_cuda [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+
+  if [ "${install_runtime}" = "1" ]; then
+    packages+=(${NV_LIBCUBLAS_PACKAGE-} ${NV_LIBNCCL_PACKAGE-} ${NV_LIBNPP_PACKAGE-})
+
+
+    packages+=(${NV_CUDA_LIB_VERSION+cuda-libraries-${cuda_version[0]}-${cuda_version[1]}-${NV_CUDA_LIB_VERSION}}
+               ${NV_NVTX_VERSION+cuda-nvtx-${cuda_version[0]}-${cuda_version[1]}-${NV_NVTX_VERSION}})
+
+    if [ "${cuda_version[0]}" = "10" ]; then
+      pcakaged+=(${NV_LIBNPP_VERSION+cuda-npp-${cuda_version[0]}-${cuda_version[1]}-${NV_LIBNPP_VERSION}})
```
[the remark this answers is not in the document]
Comment (andyneff) [reply 30_install_cuda:70]:
```suggestion
      packages+=(${NV_LIBNPP_VERSION+cuda-npp-${cuda_version[0]}-${cuda_version[1]}-${NV_LIBNPP_VERSION}})
```

**Review by andyneff — COMMENTED** [review]
Review 1 [skip ci]

File: README.rst (hunk 12 above)
Comment (andyneff) [inline README.rst:366]:
```suggestion
- Setting certain environment variables at container create time
- Setting various environment variables at container run time
```

File: recipe_cudagl.Dockerfile (hunk 10 above)
Comment (andyneff) [reply recipe_cudagl.Dockerfile:125]:
```suggestion
              echo 'PKG_CONFIG_PATH=${PKG_CONFIG_PATH-}${PKG_CONFIG_PATH:+:}/usr/local/lib64/pkgconfig:/usr/local/lib/pkgconfig' > /usr/local/share/just/info/cuda/30_pkgconfig_rhel; \
              ;; \
```

File: recipe_cuda.Dockerfile [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+ADD https://gitlab.com/api/v4/projects/2330984/repository/branches/master /cuda-master.json
+
+ARG CUDA_REPO_REF=5bc8c5483115b8e9c68d4f1280acb8d56196d681
+RUN if [ -f "/cuda-master.json" ]; then \
+      new_master_ref="$(sed 's|.*"id":"\([^"]*\).*|\1|' /cuda-master.json)"; \
+      rm /cuda-master.json; \
+      if [ "${new_master_ref}" != "${CUDA_REPO_REF}" ]; then \
+        mkdir -p /usr/local/share/just/user_run_patch/; \
+        echo "echo 'Cuda recipe is out of date. Consdier PRing to update CUDA_REPO_REF' >&2" > /usr/local/share/just/user_run_patch/00_cuda_outdated_warning; \
+        echo "echo 'Consider updating CUDA_REPO_REF(${CUDA_REPO_REF}) to ${new_master_ref} in a PR' >&2" >> /usr/local/share/just/user_run_patch/00_cuda_outdated_warning; \
```
Comment (andyneff) [inline recipe_cuda.Dockerfile:18]:
```suggestion
        echo "echo 'Cuda recipe is out of date.' >&2" > /usr/local/share/just/user_run_patch/00_cuda_outdated_warning; \
        echo "echo 'Consider submitting a PR to this repo to updats CUDA_REPO_REF(${CUDA_REPO_REF}) to ${new_master_ref}' >&2" >> /usr/local/share/just/user_run_patch/00_cuda_outdated_warning; \
```

File: README.rst (hunk 3 above)
Comment (scott-vsi) [reply README.rst:346]: Then should it say The ref of the CUDA container repo

File: README.rst (hunk 14 above)
Comment (andyneff) [reply README.rst:408]: You're thinking runtime patches, these are build time. That line is needed

File: README.rst (hunk 3 above)
Comment (andyneff) [reply README.rst:346]:
```suggestion
Build Args                  ``CUDA_REPO_REF`` - The ref of the CUDA container repo parsed
```

File: 30_install_cuda [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+  packages+=(${NV_CUDA_COMPAT_PACKAGE-cuda-compat-${cuda_version[0]}-${cuda_version[1]}}
+             cuda-cudart-${cuda_version[0]}-${cuda_version[1]}=${NV_CUDA_CUDART_VERSION})
+
+  ###########
+  # Runtime #
+  ###########
+
+  if [ "${cuda_version[0]}" = "10" ]; then
+    packages+=(${NV_LIBCUSPARSE_VERSION+cuda-cusparse-${cuda_version[0]}-${cuda_version[1]}=${NV_LIBCUSPARSE_VERSION}})
+  else
```
[the remark this answers is not in the document]
Comment (andyneff) [reply 30_install_cuda:150]: Actually, I think only 11.0 has this package, because how NV_LIBCUSPARSE_VERSION was only used in that version

**Review by scott-vsi — APPROVED** [review]

**Pull request merged**